### PR TITLE
[chromium-base] Add new port

### LINKED ISSUE #8850: Add chromium base port
Please add port for `Chromium Base`, thank goodness. 
URL at:
```
https://chromium.googlesource.com/chromium/src/base/
```
It doesn't supported for cmake, i try to write a CMakeLists.txt for it, but always failed, SOS!

各位大大请支持一下Chrome Base库，我试过自己写cmake文件，可总是失败，那个gyp文件不好读，总是翻译失败，想把chromium base作为基础库，用cmake链接

### PR DESCRIPTION
**Pull request by vejmartin**
Follow-up from https://github.com/microsoft/vcpkg/pull/12280

- What does your PR fix?  
Fixes #8850.

- Which triplets are supported/not supported? 
`x64-osx`, `x64-windows`, `x64-windows-static`

- Have you updated the CI baseline? 
Yes

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
Yes

### CODE REVIEW

**Comment by JackBoosY** [comment]
For `x64-osx`, the ci test reported:
```
ERROR at //build/config/mac/mac_sdk.gni:89:19: Script returned non-zero exit code.
_mac_sdk_result = exec_script(script_name, sdk_info_args, "scope")
                  ^----------
Current dir: /Volumes/data/buildtrees/chromium-base/x64-osx-dbg/
Command: python /Volumes/data/buildtrees/chromium-base/src/25ce732/build/config/mac/sdk_info.py macosx
Returned 1.
stderr:

xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance
Traceback (most recent call last):
  File "/Volumes/data/buildtrees/chromium-base/src/25ce732/build/config/mac/sdk_info.py", line 152, in <module>
    FillXcodeVersion(settings, args.developer_dir)
  File "/Volumes/data/buildtrees/chromium-base/src/25ce732/build/config/mac/sdk_info.py", line 64, in FillXcodeVersion
    lines = subprocess.check_output(['xcodebuild', '-version']).splitlines()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 223, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command '['xcodebuild', '-version']' returned non-zero exit status 1

See //build/toolchain/mac/BUILD.gn:15:1: whence it was imported.
import("//build/config/mac/mac_sdk.gni")
^--------------------------------------
See //BUILD.gn:1:1: which caused the file to be included.
static_library("chromium-base") {
^--------------------------------
```
Does this port not support osx currently?

File: scripts/ci.baseline.txt [hunk 1]
```diff
@@ -222,6 +222,13 @@ chipmunk:x64-osx=fail
 chipmunk:x64-uwp=fail
 chmlib:arm-uwp=fail
 chmlib:x64-uwp=fail
+chromium-base:arm64-windows=fail
```
Comment (JackBoosY) [inline scripts/ci.baseline.txt:225]: We can use `Supports` in _CONTROL_ to tell the pipeline which triples this port supports, so we don't need to add code here.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by ras0219-msft** [comment]
An initial skim leaves me very impressed, this is an awesome PR! Thank you for avoiding `depot_tools` :)

I would like to avoid checking in any `// Copyright 2006, Google Inc.` code if possible, specifically `gtest_prod.h` and minimizing `build_overrides/build.gni` as well.

Finally, it would be very nice if `vcpkg_find_acquire_program(CLANG)` detects the copy of clang/llvm in VS2019 -- I don't know that it doesn't, but this is something I'd like to check.

**Comment by Voskrese** [comment]
@ras0219-msft 

example

`msbuild /p:LLVMInstallDir="C:\Program Files (x86)\LLVM" /p:PlatformToolset="ClangCL"`


```
vcpkg_find_acquire_program(CLANG)
get_filename_component(CLANG_DIR "${CLANG}" DIRECTORY)

msbuild /p:LLVMInstallDir="${CLANG_DIR}" /p:PlatformToolset="ClangCL"
```

**Comment by vejmartin** [comment]
Thank you for the review! 

I patched `base` so that it doesn't include `gtest_prod.h` and slimmed down `build_overrides`.
`LLVM` in `Visual Studio 2019` (including the `LLVMInstallDir` override) should now also be supported as per the [docs](https://docs.microsoft.com/en-us/cpp/build/clang-support-msbuild?view=vs-2019#configure-a-linux-project-to-use-clang-tools) :)

**Comment by Voskrese** [comment]
@vejmartin 

why not add the x86 version right away ? it will not hurt, but in the future it may be useful
I'm only talking about the script `vcpkg_find_acquire_program.cmake ` , not about the chromium-base port
https://github.com/llvm/llvm-project/releases/download/llvmorg-10.0.0/LLVM-10.0.0-win32.exe

arm64
https://github.com/llvm/llvm-project/releases/download/llvmorg-10.0.0/clang+llvm-10.0.0-aarch64-linux-gnu.tar.xz

arm
https://github.com/llvm/llvm-project/releases/download/llvmorg-10.0.0/clang+llvm-10.0.0-armv7a-linux-gnueabihf.tar.xz

**Comment by vejmartin** [comment]
@Voskrese, the 32-bit version of LLVM should now get selected if the host is 32-bit Windows.

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
Another big step, thanks for this PR!

**Comment by iHuahua** [comment]
什么时候合进master啊？

**Comment by iHuahua** [comment]
@vejmartin does it supported linux?

**Comment by vejmartin** [comment]
@iHuahua, this port only adds support for macOS and Windows (x64) as these are the platforms I can debug locally. I'd expect adding support for Linux would be similar in approach.

**Comment by iHuahua** [comment]
when I call `#include <base/files/file_util.h>`, it has build failed with 
```
[build] ../vcpkg/installed/x64-osx/include/chromium-base/base/template_util.h:89:78: error: a space is required between consecutive right angle brackets (use '> >')
[build]                    void_t<typename std::iterator_traits<T>::iterator_category>>
[build]                                                                              ^
[build] ../vcpkg/installed/x64-osx/include/chromium-base/base/template_util.h:128:10: error: unknown type name 'constexpr'
[build]   static constexpr bool value =
[build]          ^
[build] ../vcpkg/installed/x64-osx/include/chromium-base/base/template_util.h:161:1: error: unknown type name 'constexpr'
[build] constexpr in_place_t in_place = {};
```
but i add `set(CMAKE_CXX_STANDARD 14)` in my CMakeLists.txt it's OK, does `chromium-baseConfig.cmake` missing it?

By the way, I debug the code with VSCode in macOS - `无法加载源“../src/25ce732/base/strings/string_piece.h”: 'SourceRequest' not supported.。`
@vejmartin pls help me, thx.

**Comment by vejmartin** [comment]
> when I call `#include <base/files/file_util.h>`, it has build failed with
> ...
> but i add `set(CMAKE_CXX_STANDARD 14)` in my CMakeLists.txt it's OK, does `chromium-baseConfig.cmake` missing it?

@iHuahua, the standard requirement is indeed missing. Thank you for pointing this out!  
As you've also found, setting the global `CXX_STANDARD` property fixes the error. This is not uncommon :)   
CMake does support cascading the requirement via `target_compile_features(target PUBLIC cxx_std_14)`, so I'll add that to `chromium-baseConfig.cmake`.

**Comment by iHuahua** [comment]
> > when I call `#include <base/files/file_util.h>`, it has build failed with
> > ...
> > but i add `set(CMAKE_CXX_STANDARD 14)` in my CMakeLists.txt it's OK, does `chromium-baseConfig.cmake` missing it?
> 
> @iHuahua, the standard requirement is indeed missing. Thank you for pointing this out!
> As you've also found, setting the global `CXX_STANDARD` property fixes the error. This is not uncommon :)
> CMake does support cascading the requirement via `target_compile_features(target PUBLIC cxx_std_14)`, so I'll add that to `chromium-baseConfig.cmake`.

@vejmartin thx for reply, can you debug `step into` chromium-base source code?  I don't known which setting I missing, it looks like source code location deos not been setted, I don't known does `ninja` write source file location info into library has removed location info or not

**Comment by vejmartin** [comment]
@iHuahua, I've added the c++14 requirement, and the debug symbols should now use absolute paths instead of relative ones as previously. Please let me know if there are any issues.

@JackBoosY, any way I can help move the discussion forward?

**Comment by iHuahua** [comment]
> @iHuahua, I've added the c++14 requirement, and the debug symbols should now use absolute paths instead of relative ones as previously. Please let me know if there are any issues.
> 
> @JackBoosY, any way I can help move the discussion forward?

@vejmartin Great, I will try again later. Then, my project had to support Android/Linux/macOS/iOS/Windows, do you have a plan to support them?

**Comment by vejmartin** [comment]
@iHuahua, I can only support macOS and Windows currently. If anybody would like to add support for the other platforms, I'd be happy to help :)

**Comment by iHuahua** [comment]
@vejmartin I have tried to support linux last night, but haven't succeed, some issue i can not fix, if you have time, pls help me.

**Comment by iHuahua** [comment]
@vejmartin hi, i have build the linux chromium-base succeed, but I got an other issue when link the library.

In `.cc` file:
```c++
#include <base/files/file_util.h>
#include <base/path_service.h>
#include <base/logging.h>

int main(int _Argc, char** _Args) {
    base::FilePath basepath;
    base::PathService::Get(base::DIR_EXE, &basepath);
    LOG(INFO) << "Exe path : " << basepath;
    return 0;
}
```

error:
```
/bin/ld: ../vcpkg/installed/x64-linux/debug/lib/libbase.a: error adding symbols: malformed archive
```

it's a same issue in macOS. the other day, I can run in macOS, but failed yesterday

**Comment by JackBoosY** [comment]
```
Installing package crashpad[core]:x64-windows-static...
The following files are already installed in D:/installed/x64-windows-static and are in conflict with crashpad:x64-windows-static

Installed by chromium-base:x64-windows-static
    debug/lib/base.lib
    lib/base.lib

Elapsed time for package crashpad:x64-windows-static: 1.169 min
```
Please move them to _lib/chromium_ directory.

File: scripts/cmake/vcpkg_find_acquire_program.cmake [hunk 2]
```diff
@@ -279,6 +280,37 @@ function(vcpkg_find_acquire_program VAR)
...
+        "${DOWNLOADS}/tools/clang/${SUBDIR}/bin")
+
+      if(DEFINED ENV{PROCESSOR_ARCHITEW6432})
+        set(HOST_ARCH_ $ENV{PROCESSOR_ARCHITEW6432})
+      else()
+        set(HOST_ARCH_ $ENV{PROCESSOR_ARCHITECTURE})
+      endif()
+      
+      if(HOST_ARCH_ MATCHES "64")
+        set(URL "https://github.com/llvm/llvm-project/releases/download/llvmorg-10.0.0/LLVM-10.0.0-win64.exe")
```
Comment (JackBoosY) [inline scripts/cmake/vcpkg_find_acquire_program.cmake:303]: I think we already have llvm here, so shouldn't we use the exist one?

File: scripts/cmake/vcpkg_install_gn.cmake [hunk 3]
```diff
@@ -14,18 +15,25 @@
...
-    cmake_parse_arguments(_vig "" "SOURCE_PATH" "TARGETS" ${ARGN})
+    cmake_parse_arguments(_vig "" "SOURCE_PATH;PATH_SUFFIX" "TARGETS" ${ARGN})
     
     if(NOT DEFINED _vig_SOURCE_PATH)
         message(FATAL_ERROR "SOURCE_PATH must be specified.")
     endif()
 
+    if(NOT DEFINED _vig_PATH_SUFFIX)
+        set(_vig_PATH_SUFFIX "")
+    endif()
```
Comment (JackBoosY) [inline scripts/cmake/vcpkg_install_gn.cmake:35]:
```suggestion
    if(NOT DEFINED _vig_PATH_SUFFIX)
        set(_vig_PATH_SUFFIX "")
    else()
        set(_vig_PATH_SUFFIX "/${_vig_PATH_SUFFIX}")
    endif()
```

File: scripts/cmake/vcpkg_find_acquire_program.cmake [hunk 4]
```diff
@@ -279,6 +280,37 @@ function(vcpkg_find_acquire_program VAR)
         set(PATHS /usr/local/opt/bison/bin)
       endif()
     endif()
+  elseif(VAR MATCHES "CLANG")
+    set(PROGNAME clang)
+    set(SUBDIR "clang-10.0.0")
+    if(CMAKE_HOST_WIN32)
+      set(PATHS 
+        # Support LLVM in Visual Studio 2019
```
Comment (JackBoosY) [inline scripts/cmake/vcpkg_find_acquire_program.cmake:288]: Does this support Visual Studio 2015 or 2017?

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: scripts/cmake/vcpkg_find_acquire_program.cmake (hunk 4 above)
Comment (vejmartin) [reply scripts/cmake/vcpkg_find_acquire_program.cmake:288]: As far as I know, LLVM support was first added in Visual Studio 2019, https://devblogs.microsoft.com/cppblog/clang-llvm-support-in-visual-studio/.

File: scripts/cmake/vcpkg_find_acquire_program.cmake (hunk 2 above)
Comment (vejmartin) [reply scripts/cmake/vcpkg_find_acquire_program.cmake:303]: Is LLVM installed elsewhere in `vcpkg`?

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
@BillyONeal Ping for merge this PR.

File: scripts/cmake/vcpkg_find_acquire_program.cmake (hunk 2 above)
Comment (BillyONeal) [reply scripts/cmake/vcpkg_find_acquire_program.cmake:306]: We install it on the CI machines but vcpkg doesn't install it. I'm not sure how I feel about that; we have tried to not get in the business of deploying the whole toolset but maybe it makes sense here....

**Comment by BillyONeal** [comment]
Tagging requires:discussion to confirm that deploying LLVM like that is OK. I'm 99.9% sure it's fine but want to be safe.

**Comment by BillyONeal** [comment]
We discussed this in person today and we're OK with the LLVM deployment. However we aren't sure about the path-suffix part https://github.com/microsoft/vcpkg/pull/12396/commits/7ddc48a3484172de24e6072e17ddaf26b617d047 . Generally speaking all libraries should be under lib or lib/manual-link, so that they can be automatically detected upon use. But base.lib is probably too generic a name. Is it named base.lib in chromium itself?

/cc @ras0219 @ras0219-msft

**Comment by BillyONeal** [comment]
Everyone loves when the upstream for the port we use in our end to end tests is down eh?

**Comment by iHuahua** [comment]
@vejmartin hi, supperman, I try to link chromium-base(`x64-windows`) with Visual Studio 2019, it crash when chromium-base generate string, such as 
```
std::cout << base::GenerateGUID();
```
And build failed on `x64-windows-static` with `error : /failifmismatch: mismatch detected for '_ITERATOR_DEBUG_LEVEL':`

**Comment by BillyONeal** [comment]
@iHuahua That means your program needs to be built with the same IDL setting `chromium-base` uses. vcpkg isn't injecting that so you might need to look upstream.

@vejmartin Thanks for your contribution!

**Comment by vejmartin** [comment]
@BillyONeal, happy to help 😄 

@iHuahua, chromium-base disables iterator debugging, so `_ITERATOR_DEBUG_LEVEL=0` needs adding to practically all libraries linked together with `chromium-base` (the ones that use the C++ Standard Library). To do that for other libraries in `vcpkg`, you could also add it to the [VCPKG_CXX_FLAGS](https://github.com/microsoft/vcpkg/blob/55be137d5ba6230f62aa7fb0994fda1728ce7fef/docs/users/triplets.md#vcpkg_cxx_flags) flags in a custom triplet.

**Pull request merged**